src/main/java/io/openliberty/website/TLSFilter.java

```diff
@@ -125,12 +136,30 @@ public void doFilter(ServletRequest req, ServletResponse resp, FilterChain chain
                 }
                 if (doGzip) {
                     response.setHeader("Content-Type", "text/html");
-                    response.setHeader("Content-Encoding", "gzip");
                     doFilter = false;
                     try {
-                        req.getRequestDispatcher(uri.concat(".gz")).include(req, response);
+                        if (queryString != null) {
+                            uriQueryString = uri +"?"+ queryString;
+                            docsPage = Jsoup.connect("https://openliberty.io"+uriQueryString).get();
+                            String ifError = docsPage.select("head > title").first().text();
+                            if(ifError.contains("404")){
+                                response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+                                response.sendRedirect("/404.html");
+                            }
+                            else {
+                                String updatedCanonicalUrl = uriQueryString.replace(uriQueryString.split("\\/")[2], "latest");
+                                Element result = docsPage.select("link[rel=\"canonical\"]").first();
+                                result.attr("href", "https://openliberty.io"+updatedCanonicalUrl);
+                                htmlResponse = response.getWriter();
+                                htmlResponse.println(docsPage.html());
+                            }
+                        }
+                        else {
+                            response.setHeader("Content-Encoding", "gzip");
+                            req.getRequestDispatcher(uri.concat(".gz")).include(req, response);
+                        }
                     }
-                    catch(FileNotFoundException e) {
+                    catch(FileNotFoundException | NullPointerException e) {
```

Comment:
Can you describe which method calls could throw a NullPointerException (NPE)? I would prefer that we check for nulls in our code and only use the catch for unexpected situations.

Reply:
If we test with not formal version like 22..0.8 it throws NullPointerException

Reply:
I would like to avoid using Java exceptions as a way to handle scenarios that we know can happen. Can you help me find the line that processes the version string? Perhaps we can use a regular expression to validate that there are four integers delimitated by periods and add error handling when the criteria is not met.

For example, in this PR you knew `docsPage.select("head > title").first().text();` can result in a "404".
```
if(ifError.contains("404")){
  response.setStatus(HttpServletResponse.SC_NOT_FOUND);
  response.sendRedirect("/404.html");
}
```